src/main/java/com/microsoft/graph/authentication/TokenCredentialAuthProvider.java

```diff
@@ -0,0 +1,141 @@
+     * @param scopes Specified desired scopes of the Auth Provider
+     * @param tokenObtentionTimeout Maximum time to wait for token obtention. Default 10 minutes. Use lower value on application with stable connectivity and no user interactions.
+     */
+    public TokenCredentialAuthProvider(@Nonnull final List<String> scopes, @Nonnull final Duration tokenObtentionTimeout, @Nonnull final TokenCredential tokenCredential) {
+        if(tokenCredential == null) {
+            throw new IllegalArgumentException("tokenCredential parameter cannot be null.");
+        }
+        if(tokenObtentionTimeout == null) {
+            throw new IllegalArgumentException("tokenObtentionTimout parameter cannot be null");
+        }
+        if(scopes == null || scopes.isEmpty()) {
+            throw new IllegalArgumentException("scopes parameter cannot be null or empty");
+        }
+        this.context = new TokenRequestContext();
+        this.context.setScopes(scopes);
+        this.tokenCredential = tokenCredential;
+        this.tokenBlockTimeout = tokenObtentionTimeout;
+    }
+
+    /**
+     * Authenticates the request
+     *
+     * @param request the request to authenticate
+     */
+    @Override
+    public void authenticateRequest(@Nonnull final IHttpRequest request) throws AuthenticationException {
+        if(request == null) {
+            throw new IllegalArgumentException("request parameter cannot be null.");
+        }
+        final URL requestUrl = request.getRequestUrl();
+        if(requestUrl != null && ShouldAuthenticateRequest(requestUrl.toString())) {
+            request.addHeader(InternalAuthConstants.AUTHORIZATION_HEADER, InternalAuthConstants.BEARER + getAccessToken());
+        }
+    }
+
+    /**
+     * Authenticates the request
+     *
+     * @param request the request to authenticate
+     * @return Request with Authorization header added to it
+     */
+    @Override
+    @Nonnull
+    public Request authenticateRequest(@Nonnull final Request request) throws AuthenticationException {
+        if(request == null) {
+            throw new IllegalArgumentException("request parameter cannot be null.");
+        }
+        final HttpUrl requestUrl = request.url();
+        if(requestUrl != null && ShouldAuthenticateRequest(requestUrl.toString())) {
+            return request.newBuilder()
+                    .addHeader(InternalAuthConstants.AUTHORIZATION_HEADER, InternalAuthConstants.BEARER + getAccessToken())
+                    .build();
+        } else {
+            return request;
+        }
+    }
+
+    private static final HashSet<String> validGraphHostNames = new HashSet<>(Arrays.asList("graph.microsoft.com", "graph.microsoft.us", "dod-graph.microsoft.us", "graph.microsoft.de", "microsoftgraph.chinacloudapi.cn"));
+    private boolean ShouldAuthenticateRequest(@Nonnull final String requestUrl) {
```

Comment:
Please use isRequestAuthentic or something indicating the verification or boolean return type.